### ISSUE TITLE
Allow more customization of transparency for GUI clients.

### ISSUE THREAD

**Issue by fredizzimo** [neovim/neovim#24472]
### Problem

Users of Neovide have wished to have more possibilities to customize the transparency, than what Neovim currently offers. See the discussion here https://github.com/neovide/neovide/pull/1790, between me an @A-Lamia for example. There was also some discussions in the Neovide discord channel with @TheSast.

I think there are two main problems at the moment.
1. The highlight group `blend` only applies to floating windows. so it's not possible to customize parts of the main ui like the main background and the status bar. This might only be a documentation issue though
> blend={integer}					highlight-blend
	Override the blend level for a highlight group within the popupmenu
	or floating windows. Only takes effect if ['pumblend'](https://neovim.io/doc/user/options.html#'pumblend') or ['winblend'](https://neovim.io/doc/user/options.html#'winblend')
	is set for the menu or window. See the help at the respective option.
2. There's no way to set blend for the foreground color. Probably the biggest use case for this is to make the statusbar fully transparent, since some of the "background" colors are often done with characters using foreground colors

### Expected behavior

The transparency is configurable for all visible elements.

**Comment by clason** [comment]
PR welcome.

**Comment by fredizzimo** [comment]
This has also been discussed here:
https://github.com/neovim/neovim/discussions/31737

I think the easiest way to implement this without introducing breaking changes is the following:
1. Introduce `blendfg` and `blendbg`, which applies to the background and the foreground color respectively
2. `blend` is deprectated for backwards compatibility and works the same way as currently, but `blendbg` always takes priority if both are specified
3. `winblend` and `pumblend` only sets the defaults for `blendbg`, when it's not specified and never toggles the blending on/off for hightlight groups that sets it
4. `blendfg` and `blendbg` applies to all windows, both floating and regular ones

NOTE. `blendbg` and `blendfg` might not be implementable in the TUI currently. But for example the [Kitty Protocol](https://sw.kovidgoyal.net/kitty/protocol-extensions/) could probably be extended with support for it. In any case it does not break anything if not supported, since it's just affects the desktop visibility through the background.

**Comment by fredizzimo** [comment]
The `highlight-blend` also has some very weird behaviour of affecting things on top of it. I'm not sure if that's on purpose or not, but I don't think the `blendbg` should work that way. For example, it can turn the visual selection transparent
![Image](https://github.com/user-attachments/assets/b903d691-6c3e-4e38-942e-6bcf3ece390f)
Or make the terminal background colors transparent
![Image](https://github.com/user-attachments/assets/e8c4545e-a15c-4fa8-9425-bde0cdfc78ed)

In both cases `winblend=1` just to enable the `blend` attribute of the custom highlight on the second line in the first example and for `FloatNormal` in the second example.

**Comment by fredizzimo** [comment]
@justinmk. I would like to work on this for the Neovim 0.12 release. Proper transparency control is something that is highly desirable by some Neovide users and has been neglected for too long with only poorly done workarounds.

What do you think about my proposal here https://github.com/neovim/neovim/issues/24472#issuecomment-2563874792? For some quirks with the current `blend` attribute, see this https://github.com/neovim/neovim/discussions/31737

**Comment by justinmk** [comment]
> What do you think about my proposal here [#24472 (comment)](https://github.com/neovim/neovim/issues/24472#issuecomment-2563874792)? 

The interface SGTM.  @bfredl @luukvbaal any major objections?

**Comment by fredizzimo** [comment]
I think we need to support a second type of `blendfg` that blends against the background of the current windows first, and then possibly through to the layer below if the background is transparent. This can be manually supported by calculating the resulting color yourself. But when using images, it gets more complex, especially if we introduce support for text on top of the image.

I'm not sure about the naming though, the only thing I can come up with right now is `blendfgwin`. But I don't think it's good.

**Comment by justinmk** [comment]
`blendfg1`, `blendfg2` seem intuitive, if multiple are needed.

**Comment by bfredl** [comment]
I'd keep `blend=15` as a _non-deprecated_ alias for setting both `blendbg=15 and `blendfg=15`. sometimes you just want some blending.

> blendfg and blendbg applies to all windows, both floating and regular ones

IMO this is already an UI-side decision and something that neovide could just do if it wants to without any change. TUI just does what the TUI is able to do, which is a starting point, not intended to limit what other UI:s can do. Documentation can be improved, but it is already meant to be informative (currently you as a user can _at least_ do this), not somehow normative _against_ external improvements.

**Comment by fredizzimo** [comment]
Also pinging @kovidgoyal, in case you want to create some additional protocols to support this. I'm not sure exactly what's needed, and what's too complex. But at least per cell control of the background color transparency seems like something that could potentially be implemented, similar to this https://sw.kovidgoyal.net/kitty/conf/#the-color-table

**Comment by fredizzimo** [comment]
> IMO this is already an UI-side decision and something that neovide could just do if it wants to without any change. TUI just does what the TUI is able to do, which is a starting point, not intended to limit what other UI:s can do. Documentation can be improved, but it is already meant to be informative (currently you as a user can _at least_ do this), not somehow normative _against_ external improvements.

I agree that it's mostly a documentation issue. It could be clear that some UIs don't support it, rather than what sounds like a hard rule. 

But, there's also an implementation issue, which might be a bug. If you currently set `blend` for the `Normal` highlight group, it's not passed on to the UI through the protocol. For other highlight groups on top-level windows. like `StatusLine` it is.

**Comment by fredizzimo** [comment]
> I'd keep `blend=15` as a _non-deprecated_ alias for setting both `blendbg=15` and `blendfg=15`. sometimes you just want some blending.

There's a subtle difference, and the reason for the deprecation, currently with 
`set winblend=0` and `hi NormalFloat guibg=red blend=50` it looks like this

![Image](https://github.com/user-attachments/assets/6dc11641-e81a-4867-8ad2-4319ecc4a410)

I would like to see that the color scheme has full control and it looks like this:

![Image](https://github.com/user-attachments/assets/d4effb47-096d-42bb-84f2-df839c0e1cdd)

**Comment by kovidgoyal** [comment]
On Thu, May 01, 2025 at 05:09:39AM -0700, fredizzimo wrote:
> fredizzimo left a comment (neovim/neovim#24472)
>
> Also pinging @kovidgoyal, in case you want to create some additional protocols to support this. I'm not sure exactly what's needed, and what's too complex. But at least per cell control of the background color transparency seems like something that could potentially be implemented, similar to this https://sw.kovidgoyal.net/kitty/conf/#the-color-table

Not likely to happen anytime soon, it would be a big performance hit,
essentially bg colors would need to be become 32bit from 24bit currently.
This has all sorts of non-local performance effects. Colors currently
are stored in 25 bits the extra bit is used to flag color table colors,
so to accommodate alpha one would need to either limit it to 7 bits or
bump up bg colors to 64 bits. And anyway IMO HDR colors (aka 10 bits per
channel) is both a more important and less invasive change.

Note that in kitty you can set and control the opacity of upto 8
background colors, but definitely not per cell. That is more than
sufficient for most use cases, but it *does not* work for foreground
colors used as backgrounds via powerline type symbols.

See https://sw.kovidgoyal.net/kitty/color-stack/#setting-and-querying-colors
look for transparent_background_colors.

In theory I could add 8 transparent_foreground_colors as well, but not
too keen. IMO powerline sections shouldn't be transparent anyway, they
show supposedly important information.

Though if one really wants too, the solution for powerline type symbols
is really to stop using text symbols for them and instead draw them as
bitmaps sized to the cell and use the graphics protocol, to place them
under the text, then you have perfect control over their transparency.

As for the general problem of blending popups and UI elements, the
solution for that is for terminals to implement native popup windows
the terminal programs wont need hacks to implement it. This is on my
TODO list and I already have a spec mostly designed for it in my head,
watch the kitty issues for an RFC when I have something more concrete.

**Comment by fredizzimo** [comment]
> Note that in kitty you can set and control the opacity of upto 8 background colors, but definitely not per cell. That is more than sufficient for most use cases, but it *does not* work for foreground colors used as backgrounds via powerline type symbols.

Perhaps there could be a way to control these colors through escape codes? That way Neovim could automatically do it in most cases. 

For the powerline, the problem is mostly that some parts of it is typically built with glyphs, so it's causing problems when using a semi-transparent background. But it could also be fixed by using the image protocol for those as you mentioned.

![Image](https://github.com/user-attachments/assets/54733c2c-fa02-4067-8ef4-589e5a52c7e8)

**Comment by kovidgoyal** [comment]
On Thu, May 01, 2025 at 06:38:21AM -0700, fredizzimo wrote:
> fredizzimo left a comment (neovim/neovim#24472)
>
> > Note that in kitty you can set and control the opacity of upto 8 background colors, but definitely not per cell. That is more than sufficient for most use cases, but it *does not* work for foreground colors used as backgrounds via powerline type symbols.
>
> Perhaps there could be a way to control these colors through escape codes? That way Neovim could automatically do it in most cases.

There already is, the link I posted above tells you about it.

**Comment by fredizzimo** [comment]
> As for the general problem of blending popups and UI elements, the
solution for that is for terminals to implement native popup windows
the terminal programs wont need hacks to implement it. This is on my
TODO list and I already have a spec mostly designed for it in my head,
watch the kitty issues for an RFC when I have something more concrete.

That will become especially useful if the Neovim TUI is implemented using the multigrid protocol, which I have heard is on the roadmap. In that case most of the advantages that Neovide has over a TUI also goes away, and it might be worth considering turning it into a terminal emulator using the same popup protocol.
